### ISSUE TITLE
GetAsync fails when specifying a null index.

### ISSUE THREAD

**Issue by niemyjski** [elastic/elasticsearch-net#1980]
We have some generic code that runs and trys to specify an index for getbyid but if it can't be found it returns null. This worked in <2.0 of the client I believe this is a bug.

Fails:
```csharp
var response = await elasticClient.GetAsync<T>(id, s => s.Index(null));


System.ArgumentException

Dispatching Get() from NEST into to Elasticsearch.NET failed
Received a request marked as GET
This endpoint accepts GET
The request might not have enough information provided to make any of these endpoints:
  - /{index=<NULL>}/{type=employee}/{id=56fed5baae14593188842ee5}


   at Nest.LowLevelDispatch.GetDispatchAsync[T](IRequest`1 p) in C:\users\russ\source\elasticsearch-net\src\Nest\_Generated\_LowLevelDispatch.generated.cs:line 932
   at Nest.ElasticClient.<Nest-IHighLevelToLowLevelDispatcher-DispatchAsync>d__24`4.MoveNext() in C:\users\russ\source\elasticsearch-net\src\Nest\ElasticClient.cs:line 75
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at Foundatio.Elasticsearch.Repositories.ElasticReadOnlyRepositoryBase`1.<GetByIdAsync>d__15.MoveNext() in C:\Code\Foundatio.Repositories\src\Elasticsearch\Repositories\ElasticReadOnlyRepositoryBase.cs:line 196
```

Works:
```csharp
var response = await elasticClient.GetAsync<T>(id);
```

**Comment by niemyjski** [comment]
If you pull down foundatio repositories project and run the tests you'll run into this.

**Comment by gmarz** [comment]
You need to specify an index when getting a document.  This exception is telling you exactly that.  If it worked in 1.x then it was wrong :).

**Comment by niemyjski** [comment]
If I pass null to the index that's the same as not specifying one.. They should work the same

**Comment by niemyjski** [comment]
I feel pretty strongly that this is a bug in the client

**Comment by gmarz** [comment]
When you don't specify an index, it's being inferred from `T`.  Passing null explicitly is overriding the inferred index.

**Comment by niemyjski** [comment]
I get this but what use case would you ever pass null in here and get a valid query?

**Comment by niemyjski** [comment]
And if you'd never get a valid query then, should we throw a argumentnullexception? This difference really threw me off guard. As a lot of tests started failing.

**Comment by gmarz** [comment]
> I get this but what use case would you ever pass null in here and get a valid query?

Exactly.  This is why we throw the exception when you pass it null.

I get your point as well, but we try to generalize this logic for all API calls within `LowLevelDispatch`, so if any required part of the path is missing we throw.  For some calls it's valid to override the inferred index and set it to null (_search for instance).

**Comment by niemyjski** [comment]
The stack trace is terrible due to how its executed :(

**Comment by gmarz** [comment]
But

```
System.ArgumentException

Dispatching Get() from NEST into to Elasticsearch.NET failed
Received a request marked as GET
This endpoint accepts GET
The request might not have enough information provided to make any of these endpoints:
  - /{index=<NULL>}/{type=employee}/{id=56fed5baae14593188842ee5}
```

Is pretty descriptive :)

**Comment by niemyjski** [comment]
Yes, it's much better than what it is but kinda of misleading until your used to all the new stack traces. I was expecting a request to be made and failed so I was looking in fiddler first ready to grab the request json.

**Comment by gmarz** [comment]
Gotcha.  Going to close this.  Please do keep opening issues though as you run into things. :+1:

**Comment by gmarz** [comment]
@niemyjski I stand corrected here :)

Discussed this further with @Mpdreamz  and @forloop and it makes sense for `.Index(null)` to behave as if it was never called.  This is a general rule the fluent API should follow everywhere.

**Comment by gmarz** [comment]
Closed via #2030
